feat: prover escrow and 712-signed quotes

just-mitch opened the pull request:
Fix https://github.com/AztecProtocol/aztec-packages/issues/8573

Introduces an escrow contract that is able to store test tokens from provers.

The implementation allows the rollup contract to stake deposited bonds, and unstake them.

Provers may initiate withdraws which are not executable until 3 epochs after their initiation.

Proposers will use the `minBalanceAtTime` function to ensure that a prover has sufficient funds at the slot they would "cash in" the quote.

Additionally, the signatures used on the epoch proof quotes have been updated to use EIP 712 style signing and verification as part of the proof claim process.

Other cleanup includes:
- Consolidating the erc20 token used in tests to TestERC20
- Fixing a bug in the publisher
- Enabling the e2e prover coordination test in ci

Future work includes:
1. Deploying the escrow along with the rollup.
2. Configuring provers to deposit funds.
3. Having the rollup actually use the escrow.

Maddiaa0 commented on l1-contracts/src/core/ProofCommitmentEscrow.sol:
```diff
@@ -0,0 +1,153 @@
+  IERC20 public token;
+  Stake public stake;
+
+  modifier onlyOwner() {
```
its more common to use the ownable library than to implement this ourselves

Maddiaa0 commented on l1-contracts/src/core/ProofCommitmentEscrow.sol:
```diff
@@ -0,0 +1,153 @@
+    _;
+  }
+
+  modifier hasBalance(address _prover, uint256 _amount) {
```
Nit: Stylewise, Im generally not a fan of using modifiers to perform anything other than an address check, I find it more clear when this check is explicitly written into a function.

Maddiaa0 commented on l1-contracts/src/core/ProofCommitmentEscrow.sol:
```diff
@@ -0,0 +1,153 @@
+    Constants.ETHEREUM_SLOT_DURATION * Constants.AZTEC_EPOCH_DURATION * 3;
+  mapping(address => uint256) public deposits;
+  mapping(address => WithdrawRequest) public withdrawRequests;
+  IERC20 public token;
```
Token can be made immutable

Maddiaa0 commented on l1-contracts/src/core/ProofCommitmentEscrow.sol:
```diff
@@ -0,0 +1,153 @@
+   */
+  function executeWithdraw() external override {
+    WithdrawRequest storage request = withdrawRequests[msg.sender];
+    if (request.executableAt > block.timestamp) {
```
>=

Maddiaa0 commented on l1-contracts/src/core/ProofCommitmentEscrow.sol:
```diff
@@ -0,0 +1,153 @@
+   * @notice Execute a mature withdrawal request
+   */
+  function executeWithdraw() external override {
+    WithdrawRequest storage request = withdrawRequests[msg.sender];
```
If you annotate this with memory, you can avoid caching amount below

Maddiaa0 commented on yarn-project/foundation/src/crypto/keccak/index.ts:
```diff
@@ -1,12 +1,17 @@
  * @returns The computed Keccak-256 hash as a Buffer.
  */
-export function keccak256(input: Buffer) {
+export function keccak256(input: Buffer | Buffer32) {
```
nice

LHerskind commented on yarn-project/circuit-types/src/prover_coordination/epoch_proof_quote_payload.ts:
```diff
@@ -50,9 +52,18 @@ export class EpochProofQuotePayload implements Signable {
+    ),
+  );
+
   getPayloadToSign(): Buffer {
```
Should be updated similar to the EIP712.

just-mitch replied:
Actually I think this one is correct.

just-mitch replied:
Let me know how you think it should change.

just-mitch replied:
Actually, since we just consult the rollup for the digest, this can simply be removed.

just-mitch committed:
Introduces an escrow contract that is able to store test tokens from provers.

The implementation allows the rollup contract to stake deposited bonds, and unstake them.

Provers may initiate withdraws which are not executable until 3 epochs after their initiation.

Proposers will use the minBalanceAtTime function to ensure that a prover has sufficient funds at the slot they would "cash in" the quote.

Additionally, the signatures used on the epoch proof quotes have been updated to use EIP 712 style signing and verification.

just-mitch commented on l1-contracts/src/core/ProofCommitmentEscrow.sol:
```diff
@@ -0,0 +1,153 @@
+    external
+    override
+    onlyOwner
+    hasBalance(_prover, _amount)
```
> overwrite the old

Yes, but that's by design. There can only be one valid proof claim at a time, and escrow trusts the rollup to only/always stakes when a valid claim comes in, and unstakes when the proof lands. `stakeBond` will overwrite the old if and only if a valid proof doesn't land.

just-mitch commented on l1-contracts/src/core/ProofCommitmentEscrow.sol:
```diff
@@ -0,0 +1,153 @@
+   *
+   * @dev     Only callable by the owner
+   */
+  function unstakeBond() external override onlyOwner {
```
The deposit amount could be removed from the `proofClaim`. The rest of the information in the claim is specific to the rollup, and not shared over here. 

Presently the escrow is deployed as a separate contract, so we would need to balance cost of duplicate storage versus calling across contracts to get data.

But I think we should discuss that elsewhere.

just-mitch commented on l1-contracts/src/core/libraries/EpochProofQuoteLib.sol:
```diff
@@ -0,0 +1,64 @@
+
+  function hash(EpochProofQuote memory quote) internal pure returns (bytes32) {
+    return keccak256(
+      abi.encode(
```
It's in the `toDigest` function lower down. I was following https://github.com/ethereum/EIPs/blob/master/assets/eip-712/Example.sol

just-mitch commented on l1-contracts/src/core/libraries/crypto/EIP712Lib.sol:
```diff
@@ -0,0 +1,27 @@
+// Copyright 2024 Aztec Labs.
+pragma solidity >=0.8.27;
+
+library EIP712Lib {
```
It had existed to have a separate, standalone utility to provde the domain hash that didn't need a deployed rollup contract.

Will remove, and inherit EIP712 directly.

just-mitch commented on l1-contracts/src/core/libraries/crypto/EIP712Lib.sol:
```diff
@@ -0,0 +1,27 @@
+  bytes32 public constant EIP712DOMAIN_TYPEHASH =
+    keccak256("EIP712Domain(string name,string version)");
+
+  bytes32 public constant DOMAIN_SEPARATOR =
```
Yes, I was thinking we could have it this way until the deployed address was certain, which would allow clients to create signatures without needing to query the contract to get the requisite info. I'm fine to remove in favor of inheriting openzeppelin's EIP712 as mentioned.

just-mitch committed:
refactor the 712 flow on the contracts

just-mitch committed:
rename the erc20 portal contract

just-mitch committed:
enable proof quote signature verification on l1
update tests

just-mitch committed:
remove dead code

just-mitch committed:
make updates to test
fix formatting

just-mitch committed:
cleanup

LHerskind commented on l1-contracts/src/core/ProofCommitmentEscrow.sol:
```diff
@@ -103,9 +97,8 @@ contract ProofCommitmentEscrow is IProofCommitmentEscrow {
    *          The prover's balance will be reduced by the bond amount
    */
-  function stakeBond(uint256 _amount, address _prover) external override onlyRollup {
+  function stakeBond(address _prover, uint256 _amount) external override onlyRollup {
```
Natspec should have the params etc. But think we are generally missing some of that. So seems like something that could be acceptable for now, and then lets create a big ass "update natspec" issue.

Created #8912

LHerskind commented on l1-contracts/test/prover-coordination/ProofCommitmentEscrow.t.sol:
```diff
@@ -13,51 +13,45 @@ import {TestERC20} from "../TestERC20.sol";
+    ESCROW = new ProofCommitmentEscrow(TOKEN, address(this));
+  }
+
   function testDeposit() public setupWithApproval(address(42), 100) {
```
Slightly annoying here, but if the input is always the `_prover` and it will be the `prover` value it don't seem useful to provide as input. Also, if always having the same inputs might as well remove them from here.

LHerskind reviewed (approved):
Really minor comment, is acceptable but think we need to cleanup natspecs and some of the modifiers since we can simplify it slightly.
